keps/sig-architecture/3784-database-of-names/README.md

```diff
@@ -0,0 +1,445 @@
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Kubernetes has many extensible fields which can carry either "standard"
+(contextual) or user-defined values.  Examples include (but are not limited to)
+label keys, annotation keys, node taints, pod resource names, conditions, etc.
+
+In most cases we define a "reserved" block, which is set aside for the project
+to use.  Labels and annotations which are prefixed with a `k8s.io` or
+`kubernetes.io` domain are reserved, and should not be used by end-users,
+whereas unprefixed names are reserved for end-users and should not be used by
+Kubernetes or any other project.
+
+We have made great use of these reserved names in myriad ways, all over the
+project.  What we have not had is a common way to know WHICH names we are
+using, what they mean, where they apply, or who is responsible for them.  In
+fact, the only way to find them is to read human-oriented documentation or to
+grep the codebase(s).
+
+This KEP proposes to store these names in a way that is searchable, machine
+parseable, but also relatively human-friendly.
+
+## Motivation
+
+We have too many KEPs and PRs that add annotation- or label-keys without any
+sort of collision detection or governance or documentation.
+
+### Goals
+
+1) Publish a list of known values.
+2) Have a place where end-users can go to find information.
+3) Have a place we can send developers to get some modicum of review of these
+names.
+
+#### Maybe later
+
+1) Add this to the KEP template
+2) Extend this DB to highly-coupled projects (e.g. containerd, crio)
+3) Extend this DB to loosely-coupled projects (e.g. GKE, EKS, AKS)
+4) Use in-code metadata to auto-generate these DB records
+
+### Non-Goals
+
+1) Completeness.  We know that getting 100% of the worms back into the can may
+   never happen.  That doesn't mean we should not try.
+2) Enforcement.  We do not intend to inject this into code-review processes.
+   We do intend to make people AWARE of it, so they can voluntarily contribute.
+
+## Proposal
+
+The "database" should live in a new git repo (proposed name:
+"kubernetes/namesdb").  In this repo we will establish a files-and-directories
+structure which allows for easy searching and "a priori" direct indexing by
+name.  We will populate it with some of the known names from the codebase, and
+then crowd-source more.  This represents the "primary index".  We will use
+OWNERS files to delegate ownership.
+
+Given that primary index, we can programmatically produce secondary indices
+which allow easy searching by other criteria (e.g. by SIG or by API resource).
+
+We can also use this data to generate docs, if we want to make the
+"description" parts of this proposal fairly robust.
+
+This repo should probably live under k/k/staging to make PRs easier to add
+names and code which uses them at the same time.
```

Comment:
nit, this paragraph may be reworded to avoid some confusion if you don't know staging syncs to the external repo from k/k

> The "database" should live in a new git repo (proposed name:
"kubernetes/namesdb").

> This repo should probably live under k/k/staging

Reply:
ACK